rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/ChargingRSpace.scala

```diff
@@ -1,38 +1,44 @@
 package coop.rchain.rholang.interpreter.storage
 
+import java.nio.ByteBuffer
+
 import cats.effect.Sync
 import cats.implicits._
+import coop.rchain.crypto.hash.Blake2b512Random
 import coop.rchain.metrics.Span
-import coop.rchain.models.TaggedContinuation.TaggedCont.ParBody
+import coop.rchain.models.TaggedContinuation.TaggedCont.{Empty, ParBody, ScalaBodyRef}
 import coop.rchain.models._
-import coop.rchain.rholang.interpreter.Runtime.{RhoISpace, RhoTuplespace}
+import coop.rchain.rholang.interpreter.Runtime.RhoTuplespace
 import coop.rchain.rholang.interpreter.accounting._
-import coop.rchain.rspace.{
-  Blake2b256Hash,
-  Checkpoint,
-  ContResult,
-  Result,
-  SoftCheckpoint,
-  internal,
-  Match => StorageMatch
-}
+import coop.rchain.rholang.interpreter.errors.BugFoundError
+import coop.rchain.rholang.interpreter.storage.ChargingRSpace.consumeId
+import coop.rchain.rspace.{ContResult, Result, Match => StorageMatch}
 
 import scala.collection.SortedSet
 
 object ChargingRSpace {
-  def storageCostConsume(
-      channels: Seq[Par],
-      patterns: Seq[BindPattern],
-      continuation: TaggedContinuation
-  ): Cost = {
-    val bodyCost = Some(continuation).collect {
-      case TaggedContinuation(ParBody(ParWithRandom(body, _))) => body.storageCost
-    }
-    channels.storageCost + patterns.storageCost + bodyCost.getOrElse(Cost(0))
+
+  private sealed trait TriggeredBy {
+    val id: Blake2b512Random
+    val persistent: Boolean
+    val channelsCount: Int
+  }
+
+  private final case class Consume(id: Blake2b512Random, persistent: Boolean, channelsCount: Int)
+      extends TriggeredBy
+
+  private final case class Produce(id: Blake2b512Random, persistent: Boolean) extends TriggeredBy {
+    override val channelsCount = 1
   }
 
-  def storageCostProduce(channel: Par, data: ListParWithRandom): Cost =
-    channel.storageCost + data.pars.storageCost
+  private def consumeId[F[_]: Sync](continuation: TaggedContinuation): F[Blake2b512Random] =
+    //TODO: Make ScalaBodyRef-s have their own random state and merge it during its COMMs
+    continuation.taggedCont match {
+      case ParBody(value) => value.randomState.pure[F]
+      case ScalaBodyRef(value) =>
+        Blake2b512Random(ByteBuffer.allocate(8).putLong(value).array()).pure[F]
```

Comment:
@pyrocto It occured to me that Scala continuations (so: all system processes) don't have their own random state, and the dispatcher doesn't do a merge when COMM-ing with them. Do you see an issue with that?

My default reaction would be to make it uniform between rholang- and scala- continuations, and make both of them have a "randomState", and have that merged on each COMM (so that COMM always means a random state merge). Do you see reasons not to do that?

Reply:
Relevant part of dispatcher code:
https://github.com/rchain/rchain/blob/9676e49488da094428e0fdf8c312f8d8f94db221/rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala#L29

Reply:
> make it uniform between rholang- and scala- continuations, and make both of them have a "randomState", and have that merged on each COMM (so that COMM always means a random state merge).

I agree.

> Do you see reasons not to do that?

On the contrary, I think uniformity is the only way you'll get consistency across validators.

Reply:
Created https://rchain.atlassian.net/browse/RCHAIN-3791